### ISSUE TITLE
Implement Multi-Objective Routing with User Preferences

### ISSUE THREAD

**Issue by 0xtechroot** [0xtechroot/cowSolver#69]
## Overview
Implement multi-objective routing that balances price, gas, latency, and risk based on user-configurable preferences.

### 1. Price Optimization
**Minimize economic loss:**
- Price impact across all hops
- Protocol fees
- Slippage
- Spread costs

### 2. Gas Optimization
**Minimize transaction costs:**
- Estimated gas consumption
- Current gas prices (EIP-1559)
- Multi-hop overhead
- Complex operation costs

### 3. Latency Optimization
**Minimize execution time:**
- Bridge finality times
- Block confirmation delays
- Relayer response times
- Network congestion

### 4. Risk Optimization
**Minimize security exposure:**
- Bridge trust assumptions
- Protocol audit status
- Historical failure rates
- Liquidity depth

### Scoring Function
```rust
struct RouteObjectives {
    price: f64,      // Economic loss
    gas: f64,        // Gas cost in USD
    latency: f64,    // Expected time in seconds
    risk: f64,       // Risk score (0-1)
}

struct UserPreferences {
    weight_price: f64,    // Default: 0.4
    weight_gas: f64,      // Default: 0.3
    weight_latency: f64,  // Default: 0.2
    weight_risk: f64,     // Default: 0.1
}

fn route_score(objectives: &RouteObjectives, prefs: &UserPreferences) -> f64 {
    prefs.weight_price * objectives.price +
    prefs.weight_gas * objectives.gas +
    prefs.weight_latency * objectives.latency +
    prefs.weight_risk * objectives.risk
}
```

#### "Cheapest" Strategy
```rust
UserPreferences {
    weight_price: 0.6,
    weight_gas: 0.3,
    weight_latency: 0.05,
    weight_risk: 0.05,
}
```

#### "Fastest" Strategy
```rust
UserPreferences {
    weight_price: 0.2,
    weight_gas: 0.1,
    weight_latency: 0.6,
    weight_risk: 0.1,
}
```

#### "Safest" Strategy
```rust
UserPreferences {
    weight_price: 0.2,
    weight_gas: 0.1,
    weight_latency: 0.1,
    weight_risk: 0.6,
}
```

#### "Balanced" Strategy
```rust
UserPreferences {
    weight_price: 0.4,
    weight_gas: 0.3,
    weight_latency: 0.2,
    weight_risk: 0.1,
}
```

#### Price Objective
```rust
fn calculate_price_objective(route: &Route) -> f64 {
    let mut total_loss = 0.0;
    
    for hop in &route.hops {
        total_loss += hop.price_impact;
        total_loss += hop.protocol_fee;
        total_loss += hop.slippage;
    }
    
    total_loss
}
```

#### Gas Objective
```rust
fn calculate_gas_objective(route: &Route, gas_price: U256) -> f64 {
    let estimated_gas = estimate_route_gas(route);
    let gas_cost_wei = estimated_gas * gas_price;
    wei_to_usd(gas_cost_wei)
}
```

#### Latency Objective
```rust
fn calculate_latency_objective(route: &Route) -> f64 {
    let mut total_time = 0.0;
    
    for hop in &route.hops {
        total_time += match hop.venue {
            Venue::AMM(_) => 12.0,  // 1 block
            Venue::Bridge(ref b) => b.expected_time_seconds(),
            Venue::Aggregator(_) => 15.0,
        };
    }
    
    total_time
}
```

#### Risk Objective
```rust
fn calculate_risk_objective(route: &Route) -> f64 {
    let mut total_risk = 0.0;
    
    for hop in &route.hops {
        total_risk += hop.venue.risk_score();
        total_risk += hop.liquidity_risk();
    }
    
    total_risk / route.hops.len() as f64
}
```

### 2. Pareto Frontier Analysis
**Find non-dominated solutions:**
- Generate multiple candidate routes
- Calculate all objectives
- Identify Pareto-optimal routes
- Present trade-offs to user

### 3. Dynamic Weight Adjustment
**Context-aware preferences:**
- Large trades → increase weight_price
- Time-sensitive → increase weight_latency
- Experimental tokens → increase weight_risk
- High gas environment → increase weight_gas

### API Design
```rust
struct RouteRequest {
    source_token: Token,
    dest_token: Token,
    amount: U256,
    preferences: UserPreferences,
    constraints: RouteConstraints,
}

struct RouteConstraints {
    max_latency: Option<Duration>,
    max_gas: Option<U256>,
    max_risk: Option<f64>,
    allowed_bridges: Option<Vec<BridgeId>>,
}

struct RouteResponse {
    recommended_route: Route,
    alternatives: Vec<Route>,
    objectives: RouteObjectives,
    trade_offs: Vec<TradeOff>,
}
```

### Trade-Off Visualization
```rust
struct TradeOff {
    description: String,
    alternative_route: Route,
    savings: f64,
    cost: String,
}

// Example:
// "Save $2.50 in gas by using a 2-minute slower route"
// "Save 30 seconds by paying $1.20 more in fees"
```

## Implementation Phases

**Phase 1:** Objective calculation
- Price impact modeling
- Gas estimation
- Latency prediction
- Risk scoring

**Phase 2:** Multi-objective optimization
- Scoring function
- Preset strategies
- Pareto frontier
- Route ranking

**Phase 3:** User preferences
- API design
- Preference storage
- Dynamic adjustment
- Constraint handling

**Phase 4:** Trade-off analysis
- Alternative generation
- Comparison metrics
- User education
- Visualization

## Performance Targets
- **<100ms** for multi-objective route calculation
- **5-10** alternative routes per request
- **95%+** user satisfaction with recommendations
- **Pareto-optimal** solutions

## Benefits
- User empowerment
- Transparent trade-offs
- Optimized for user goals
- Competitive differentiation

## Priority
**High** - Key user experience differentiator

## Labels
`enhancement`, `routing`, `user-experience`, `research`
